cmake/Utils.cmake

```diff
@@ -278,3 +278,66 @@ function(parse_cuda_version CUDA_VERSION CUDA_VERSION_MAJOR_VAR CUDA_VERSION_MIN
 
   message(STATUS "CUDA version: ${CUDA_VERSION}, major: ${${CUDA_VERSION_MAJOR_VAR}}, minor: ${${CUDA_VERSION_MINOR_VAR}}, patch: ${${CUDA_VERSION_PATCH_VAR}}, short: ${${CUDA_VERSION_SHORT_VAR}}, digit-only: ${${CUDA_VERSION_SHORT_DIGIT_ONLY_VAR}}")
 endfunction()
+
+# Build a .so library variant for each python version provided in PYTHON_VERSIONS variable
+# if it is accesible during the build time. The library sufix is provided and specific for each python version
+#
+# supported options:
+# TARGET_NAME - umbrela target name used for this build
+# OUTPUT_NAME - library name used for this build
+# PREFIX - library prefix, if no provided there library will be named as ${TARGET_NAME}.python_specific_extension
+# OUTPUT_DIR - ouptut directory of the build library
+# PUBLIC_LIBS - list of libraries that should be linked in as a public one
+# PRIV_LIBS - list of libraries that should be linked in as a private one
+# SRC - list of source code files
+function(build_per_python_lib)
+    set(oneValueArgs TARGET_NAME OUTPUT_NAME OUTPUT_DIR PREFIX)
+    set(multiValueArgs PRIV_LIBS PUBLIC_LIBS SRC)
+    cmake_parse_arguments(PARSE_ARGV 0 PYTHON_LIB_ARG "${options}" "${oneValueArgs}" "${multiValueArgs}")
+
+    add_custom_target(${PYTHON_LIB_ARG_TARGET_NAME} ALL)
+
+    foreach(PYVER ${PYTHON_VERSIONS})
+
+        # check if listed python versions are accesible
+        execute_process(
+            COMMAND python${PYVER} -c "print('test', end='')"
+            OUTPUT_VARIABLE PYTHON_EXISTS)
+
+        if ("${PYTHON_EXISTS}" STREQUAL "test")
+            execute_process(
+                COMMAND python${PYVER}-config --extension-suffix
+                OUTPUT_VARIABLE PYTHON_SUFIX)
+            # remove newline and the extension
+            string(REPLACE ".so\n" "" PYTHON_SUFIX ${PYTHON_SUFIX})
+
+            execute_process(
+                COMMAND python${PYVER}-config --includes
+                OUTPUT_VARIABLE PYTHON_INCLUDES)
+            # split and make it a list
+            string(REPLACE "-I" "" PYTHON_INCLUDES ${PYTHON_INCLUDES})
+            string(REPLACE "\n" "" PYTHON_INCLUDES ${PYTHON_INCLUDES})
+            separate_arguments(PYTHON_INCLUDES)
+
+            add_library(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} SHARED ${PYTHON_LIB_ARG_SRC})
+            target_link_libraries(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} PUBLIC ${PYTHON_LIB_ARG_PUBLIC_LIBS})
+            target_link_libraries(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} PRIVATE ${PYTHON_LIB_ARG_PRIV_LIBS})
+            target_link_libraries(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} PRIVATE "-Wl,--exclude-libs,${exclude_libs}")
```

Comment:
Apart from list of python version, this is a second variable that is magically taken from outer scope and not as a parameter. Do you think it's worth adjusting?

Reply:
Done